### PR TITLE
[[New Exercise]] RNA Transcription

### PR DESCRIPTION
**Pull request by WammKD**
I think I took care of all required files but let me know if I missed anything.

Added the exercise to the config.json file and added the required files (an example under the corresponding directory and module file, file for writing the exercise, and the file containing tests under the source directory) along with updating the README.md.

### CODE REVIEW

File: exercises/rna-transcription/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# RNA Transcription
```
Comment (petertseng) [inline exercises/rna-transcription/README.md]: you are right that this should be RNA transcription, though the generator says "Rna Transcription" instead. That's a bit annoying discrepancy, would like to figure out how to change that so that the generator doesn't try to change it back when run on this track.

No change needs to happen for this PR, but I need to note it down for later use.

File: exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import ceylon.test { ... }
+
+// Tests adapted from x-common version 1.0.0
```
Comment (petertseng) [inline exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: given that this claims to follow x-common 1.0.0, now I am obligated to ask: is the omission of https://github.com/exercism/problem-specifications/blob/master/exercises/rna-transcription/canonical-data.json#L48-L51 intentional? Please explain why if so?

Comment (petertseng) [inline exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: I guess x-common should be changed to problem-specifications, but doesn't need to happen now

Comment (petertseng) [inline exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: given that problem-specifications is currently at 1.0.1, why not say adapted from 1.0.1?

File: exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  // strand with invalid nucleotides
+  ["XCGFGGTDTTAA",        Error()],
+  // strand with invalid nucleotides after the first nucleotide
+  ["ACGTFGTBTEAA",        Error()]
+};
+
+test
+parameters(`value cases`)
+void testHamming(String dna, String|Error expected) {
+  if(is String expected) {
```
Comment (petertseng) [inline exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: I have no ground to stand on for this request other than "all other exercises do it this way", but I've had a space between `if` and the subsequent parenthesis.

File: exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    assertEquals(transcription(dna), expected);
+  } else if(!is Error result = transcription(dna)) {
```
Comment (petertseng) [inline exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: I have no ground to stand on for this request other than "all other exercises do it this way", but I've had a space between `if` and the subsequent parenthesis.

File: exercises/rna-transcription/example/RNA.ceylon [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class Error() {}
+
+String|Error transcription(String dna) {
+  return dna.fold("" of String|Error)((String|Error r,
+                                       Character    c) =>   if(is Error r)
```
Comment (petertseng) [inline exercises/rna-transcription/example/RNA.ceylon]: I have no ground to stand on for this request other than "all other exercises do it this way", but I've had a space between `if` and the subsequent parenthesis.

Comment (petertseng) [inline exercises/rna-transcription/example/RNA.ceylon]: the `if` seems to be indented farther to the right than the `then` and `else`. Is there a reason for that?

File: exercises/rna-transcription/example/RNA.ceylon [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+                                                          then(r)
```
Comment (petertseng) [inline exercises/rna-transcription/example/RNA.ceylon]: If the parentheses around `r` were to be removed, is it still valid Ceylon? If it is still valid, shall they be removed? I don't really have ground for this request other than "I like to avoid unnecessary parentheses)

File: exercises/rna-transcription/example/RNA.ceylon [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+                                                          else(switch(c)
```
Comment (petertseng) [inline exercises/rna-transcription/example/RNA.ceylon]: If the parentheses around the entire `switch` were to be removed, is it still valid Ceylon? If it is still valid, shall they be removed? I don't really have ground for this request other than "I like to avoid unnecessary parentheses)

File: exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  // single G nucleotide strand
+  ["G",                       "C"],
+  // single A nucleotide strand
+  ["A",                       "U"],
+  // single T nucleotide strand
+  ["T",                       "A"],
+  // longer strand
+  ["ACGTGGTCTTAA", "UGCACCAGAAUU"],
+  // strand with invalid nucleotides
+  ["XCGFGGTDTTAA",        Error()],
```
Comment (petertseng) [inline exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: an unfortunate consequence of saying `Error()` here (instead of some other indication that something of type `Error` is expected) is that now I think this makes it impossible to have an `Error` that contains additional information, which students might like to do (for example `Error(InvalidNucleotide('X'))`) ?

Ah, unless the student makes a `InvalidNucleotide` subclass of `Error` and allows the `InvalidNucleotide` to contain the character. If that is possible (I think it is, but I've forgotten), then no objection.

File: exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import ceylon.test { ... }
+
+// Tests adapted from x-common version 1.0.0
+{[String, String|Error]*} cases => {
+  // empty strand
```
Comment (petertseng) [inline exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: this case is not present in https://github.com/exercism/problem-specifications/blob/master/exercises/rna-transcription/canonical-data.json#L48-L51 . shall it be added to canonical-data.json?

File: exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon (hunk 2 above)
Comment (WammKD) [inline exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: Honestly, I just hadn't been aware of that file's existence and hadn't seen any documentation on it (though, admittedly, I'd just targeted briefly the pieces of documentation that seemed relevant to what I was doing since it's somewhat spread out (to me) and the elements present in other exercises seemed straightforward).

I got the base for the test file from the Hamming exercise and just edited the parts I understood (and allowed for things to pass when I ran the tests).

In short, no reason; I'll add it in.

Comment (WammKD) [inline exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: Basically the reason in my answer above; I'll up the version number.

File: exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon (hunk 9 above)
Comment (WammKD) [inline exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: I accumulated tests from both of the Scheme and Java tracks (though I can't remember which this one came from; Java, I think). I can open up a pull request to add that test.

File: exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon (hunk 8 above)
Comment (WammKD) [inline exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: That's true; I always figured the purpose of providing the dummy class was more of a restriction as part of the solution for the problem.

I can adjust the tests to more similarly match Hamming's¸ if you'd prefer.

File: exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon (hunk 3 above)
Comment (WammKD) [inline exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: Haha; there's a style checker at work that insists of having spaces between the if and parenthesis and it makes me _so_ angry, every time. But I have no qualms with changing it. It's community code and understandable to have it be consistent. I'll get it altered.

File: exercises/rna-transcription/example/RNA.ceylon (hunk 6 above)
Comment (WammKD) [inline exercises/rna-transcription/example/RNA.ceylon]: It is valid but I feel like it breaks the consistency with the parentheses on the if and else branches of the if statement. If you'd prefer them removed, I can do that.

File: exercises/rna-transcription/example/RNA.ceylon (hunk 7 above)
Comment (WammKD) [inline exercises/rna-transcription/example/RNA.ceylon]: I would have thought it would be valid (why not the `switch` but you can remove the parentheses on the `else`?) but it isn't.

File: exercises/rna-transcription/example/RNA.ceylon (hunk 5 above)
Comment (WammKD) [inline exercises/rna-transcription/example/RNA.ceylon]: Basically, it just aligns the first parenthesis with the first parentheses of the `else` and `then`. It makes it easier for the eye to see what the arguments are for each branch of the `if` statement, to me. I can change it, if you'd prefer.

File: exercises/rna-transcription/example/RNA.ceylon [hunk 10]
```diff
@@ -2,12 +2,12 @@ class Error() {}
...
-                                                          then(r)
-                                                          else(switch(c)
-                                                               case('G') r + "C"
-                                                               case('C') r + "G"
-                                                               case('T') r + "A"
-                                                               case('A') r + "U"
-                                                               else(Error())));
+                                       Character    c) =>   if (is Error r)
+                                                          then (r)
+                                                          else (switch(c)
```
Comment (petertseng) [inline exercises/rna-transcription/example/RNA.ceylon]: Applicable to all these six lines (the `switch`, `case`, `else`) - I guess now I wonder whether there should be spaces between the keyword and the following parenthesis for these as well. It looks a little inconsistent that I asked for spaces after the `if` and `then` and the `else` on this line, but not for these six lines.

I just checked and found that https://github.com/ceylon/ceylon.formatter exists; maybe it should be used instead of me making up rules since it's not like I know what I'm doing.

I'll take a look to see whether it's usable. If it is not usable, I am starting to think it will be a bit difficult to enforce consistency across the entire project, though consistency within one file is most certainly good. I will be happy with consistently a space or consistently no space in this file.

**Comment by petertseng** [comment]
> I just checked and found that https://github.com/ceylon/ceylon.formatter exists; maybe it should be used instead of me making up rules since it's not like I know what I'm doing.

Well... it *is* runnable (in fact, sdkman installs it with Ceylon, though I have to ask for Ceylon 1.3.2 instead of 1.3.3 right now). But it formats the file in a way that I would say doesn't look that good.

```ceylon
diff --git i/exercises/rna-transcription/example/RNA.ceylon w/exercises/rna-transcription/example/RNA.ceylon
index 21dbbd9..45961e1 100644
--- i/exercises/rna-transcription/example/RNA.ceylon
+++ w/exercises/rna-transcription/example/RNA.ceylon
@@ -1,13 +1,13 @@
 class Error() {}
 
 String|Error transcription(String dna) {
-  return dna.fold("" of String|Error)((String|Error r,
-                                       Character    c) =>   if (is Error r)
-                                                          then (r)
-                                                          else (switch(c)
-                                                                case('G') r + "C"
-                                                                case('C') r + "G"
-                                                                case('T') r + "A"
-                                                                case('A') r + "U"
-                                                                else(Error())));
+    return dna.fold("" of String|Error)((String|Error r,
+            Character c) => if (is Error r)
+            then (r)
+            else (switch (c)
+                        case ('G') r + "C"
+                        case ('C') r + "G"
+                        case ('T') r + "A"
+                        case ('A') r + "U"
+                        else (Error())));
 }
```

so I'm not 100% set on using it

File: exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon [hunk 11]
```diff
@@ -1,7 +1,7 @@
 import ceylon.test { ... }
 
-// Tests adapted from x-common version 1.0.0
-{[String, String|Error]*} cases => {
+// Tests adapted from problem-specifications version 1.0.1
+{[String, String|Null]*} cases => {
```
[the remark this answers is not in the document]
Comment (WammKD) [reply exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: I don't really have a preference, either. At this point, I'll probably just leave it.

File: exercises/rna-transcription/example/RNA.ceylon (hunk 10 above)
Comment (WammKD) [reply exercises/rna-transcription/example/RNA.ceylon]: Yeah, I'd probably want the spaces if there's going to be spaces for the `if`, `then`, and `else`. Consistency is pleasing.

**Comment by WammKD** [comment]
Oh, dear. Yeah, I'm inclined towards your opinion – that doesn't look good, at all.

Sorry for the long lapse. I squashed all of the commits into one and I think I've gotten everything so this branch good to go, I believe.

File: exercises/rna-transcription/README.md (hunk 1 above)
Comment (petertseng) [reply exercises/rna-transcription/README.md]: I noted it at https://github.com/exercism/configlet/issues/76

**Comment by petertseng** [comment]
Thanks!

File: exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon (hunk 9 above)
Comment (petertseng) [reply exercises/rna-transcription/source/rnatranscription/RNAtest.ceylon]: has now been added in https://github.com/exercism/ceylon/pull/73 because https://github.com/exercism/problem-specifications/pull/1266
